### PR TITLE
Add support for a pending event to be observed immediately

### PR DESCRIPTION
**Pull request by fergusonm**
For your consideration...

When a `LiveEvent` has a value set but there are no observers the current value is simply not observed when an observer is added.  Observers will only see new values as they get set.  This is contrary to how `LiveData` behaves where a previously set value will be immediately observed by an observer. 

This can happen if, for example, a `ViewModel` sets a `LiveEvent` value during initialization and observers do not start observing until afterwards, say in `onResume`.

I feel `LiveEvent` should behave as similarly as possible to `LiveData` but with the only restriction of only emitting the value once for the current set of observers.

### CODE REVIEW

**Comment by hadilq** [comment]
Thank you for your contribution. I'm totally happy to see your PR and think it's a valid one, but if you could answer me one question correctly! Imagine multiple fragments in a page with one `ViewModel` then we trigger an event in a `LiveEvent` before the fragments are subscribing to it. Then which one of the fragments would observe the event? In your solution the first one to observe the `LiveEvent` would receive the event, but is it a correct behaviour?!

**Comment by hadilq** [comment]
I can answer my question by saying both the new `Observer`s in those fragments must receive the pending event. Soooo would you add a commit and replace `pending.getAndSet(false)` with `pending.get()`? Then all the new `Observer`s would be notified that there is an initial value for them and the `setValue` method would be responsible to change the state only. If you change it this way or find another to answer the question, I'll go with your solution. Thanks.

**Comment by fergusonm** [comment]
Hmmm... It seems to me that replacing `pending.getAndSet(false)` with just `pending.get()` causes `LiveEvent `to effectively behave the same as `LiveData` with `Transformations.distinctUntilChanged(...)`.  

This is especially noticeable on configuration change but is also observable if you remove an observer and then re-observe the same `LiveEvent`.  (To be clear I'm primarily concerned with configuration change.)  The event gets re-observed again as new observers attach to the `LiveEvent`.  This violates the principle that `LiveEvent `should only emit once. New observers see existing data and act on it, even though they had just done so prior to configuration change.

I do see your point.  However, I cannot see a way to have both the restriction of delivering the event _only once_ to the current set of observers but at the same time preserving the pending event for yet-to-be subscribed observers.

**Comment by fergusonm** [comment]
I would also like to add that even with a shared view model between fragments, any observe added before `onResume` would receive the event.  The only issue we have is with really late observers being added after one observer has observed the value.  That is if they start observing the `LiveEvent` well after `onResume` has hit.

**Comment by hadilq** [comment]
Now, it doesn't make sense! Please re-think about the logic. For instance:

> I cannot see a way to have both the restriction of delivering the event only once to the current set of observers but at the same time preserving the pending event for yet-to-be subscribed observers.

Yes! That's why you have two branches of `if` in your `setValue`, where exactly separated them. I didn't ask for that, though.

**Comment by fergusonm** [comment]
Maybe I'm confused then.  To be clear, you're suggesting I should only read the pending value when an observer is added and only set or clear the pending flag in `setValue`?

**Comment by fergusonm** [comment]
If so then if you only set a value once, before observers are present, it will forever be pending and will re-emit on configuration change.

**Comment by hadilq** [comment]
> If so then if you only set a value once, before observers are present, it will forever be pending and will re-emit on configuration change.

This is the behaviour of `LiveData` itself and we support it. Here we just guarantee that every observer would receive the data once, which makes it an event. And this is the only difference between `LiveEvent` and the `LiveData`. I may need to add this statement to the README file! Would you add this too? :D

**Comment by fergusonm** [comment]
I would disagree.  I think that the changes you are requesting turn `LiveEvent` essentially back into `LiveData` wrapped in `Transformations.distinctUntilChanged()` or the equivalent extension function.  I don't think that makes sense.  I think I'm going to maintain my fork true to the original read me.  That is true to:

> This class is extended form `LiveData` class, from androidx.lifecycle:lifecycle-extensions library, to propagate the data as an event, which means it emits data just once.

I think emitting just once is a great idea.  It solves a problem that would otherwise have to be solved by an RxJava `PublishSubject` or something else just as verbose.  

Thanks for your original idea and implementation of `LiveEvent` but I don't agree with your suggested changes.

**Comment by hadilq** [comment]
Thank you too. I believe as a rule, the behaviour of `LiveEvent` on notifying observers must be the same, for instance consider the "Two fragment scenario", which is not happening in your PR. You may agree that this rule is working for `LiveData` and `Subject`s in `Rx` library and I want to support it too. Let's find a solution that doesn't discard this rule.

**Comment by fergusonm** [comment]
I had, obviously, thought the intent of `LiveEvent` was to discard the rule of new observers seeing the existing event that `LiveData` has.  I suppose my PR expanded on that rule a bit to add initial observation rules.  To be more specific, I had though the purpose was to deliver an event once and once only to the current set of observers.  New observers would have to wait for a new value to be set.

Based on our discussion and the clarified requirements I don't see a need for `LiveEvent` as you envision it.  It seems to me that all the goals of `LiveEvent` are satisfied by existing tools within the androidx lifecycle toolkit.

The "two fragment scenario", single event delivery per observer, delivery of existing events to new observers (due to configuration change or otherwise doesn't really matter) is easily satisfied by using _only_ `LiveData` and its associated `Transformations` tools.

Maybe I'm missing something.  (And I'm hoping I am.)  Given this discussion and the clarified requirement, can you describe a use-case that `LiveEvent` satisfies that `Transformations.distinctUntilChanged(liveData)` or even just `LiveData` by itself does not?

**Comment by hadilq** [comment]
Thank you for your following up.

First the `Transformations.distinctUntilChanged` method. The document says "Creates a new LiveData object that does not emit a value until the source LiveData value has been changed. The value is considered changed if equals() yields false". So to use it, you have to implement the `equals` method, which is not required for `LiveEvent`. Even in case of "data class" of Kotlin, the developer must be careful with that method to use `Transformations.distinctUntilChanged` method. Also the use case of `LiveEvent` is different while every event would be emitted, except the initial ones, but in `distinctUntilChanged` method any data could be blocked if it equals to the previous one.

> The "two fragment scenario", single event delivery per observer, delivery of existing events to new observers (due to configuration change or otherwise doesn't really matter) is easily satisfied by using only LiveData and its associated Transformations tools.

Would be nice to explain it more? How would be the implementation? Thank you so much.

**Comment by fergusonm** [comment]
Oh wow.  I think the confusion is coming from the fact that `LiveEvent` does indeed emit to an observer more than once.  That seems like a bug based on what you mentioned.

The reason is because the wrapped observer sets the pending flat to true on `newValue`.  Every time a `LiveEvent` gets a new value sent to it, it notifies the observer and the observer's `newValue` method gets hit, which in turn resets the pending flag and the observer's `onChanged` method is called, it sees pending as true and emits the value to the observer.

I'm starting to understand why we're so out of sync now. My PR was based on how it actually behaved rather than how you intended it to.  

We have fundamentally different goals.  Correct me if I'm wrong.  You want an observer that can only see the event once.  Effectively a LiveData that can only ever have one value set on it.  All subsequent values are ignored? 

I want something more like an Rx `PublishSubject` but with the benefits of being lifecycle aware like `LiveData` is.

**Comment by amatkivskiy** [comment]
@hadilq  @fergusonm Seems I need this feature too 😞  
I am using [Cicerone](https://github.com/terrakok/Cicerone) for navigation (more precisely Single Activity container and navigating through Fragments with **Cicerone**). 

And I have a case when I am navigating back to some fragment that is already in the stack. This fragment's view is getting recreated and it resubscribes to `LiveEvent`. The problem is that I have some logic that gets executed **during** this navigation and submits events **during** navigation. And with the current implementation, all that events are ignored 😥 

@fergusonm have you tried the solution that you implemented in this PR?

**Comment by fergusonm** [comment]
Yeah I used my variation of it for a while.  It worked fine.  I'm going to be honest though, after a lot of experimentation anything that uses `LiveData` or any type of single value data holder as a backing mechanism is not an appropriate solution to transmit events between the view model and a view. 

I **do not** recommend the using `LiveEvent` as an event communication mechanism. It's simply not safe. Back to back events step on each other if there are no observers and as you are noticing events emitted before there are any observers are dropped.  

What is really needed is a stream, not a data holder.

In the time since I wrote this PR I have become much more proficient in Kotlin and coroutines. My view models now have three elements to them for outbound communication.  

For **view state** I'm using `StateFlow`.  This is a conflated stream that essentially acts like `LiveData`. It's great when you don't need to worry about intermediate values stepping on each other.  You'll only ever observe the last value emitted.

For **events** where I require guaranteed delivery I use a normal Kotlin channel that I receive as flow.  Events cannot step on each other and back to back values are buffered if there no observers, as there might be during a configuration change.  But it is backed by a channel so there are so caveats like only having a single observer. For an event system I actually think that's perfect.
eg.
```
private val _events = Channel<Event>(Channel.BUFFERED)
val events = _events.receiveAsFlow()
```

For **lossy events or notices** (I'm not quite sure the term to use) I use `SharedFlow`. If you really want multiple observers and you acknowledge that events can be lost between the time you start the first observer observing and the time the second observer observes you can use `SharedFlow`. This flow is hot and will drop values if there are no observers but once an observer attaches it broadcasts to all observers when values are emitted.  (It's great for streams like GPS location where dropped values are ok.)

Don't take my word for it. Roman Elizarov (of Jetbrains and Kotlin fame) has got a great write up on those three use cases (though only from a pure Kotlin view point and not from an Android view point).  See [https://elizarov.medium.com/shared-flows-broadcast-channels-899b675e805c](https://elizarov.medium.com/shared-flows-broadcast-channels-899b675e805c) Of particular interest is the section titled "A use-case for channels".

TL;DR don't use `LiveEvent`.  Use normal Kotlin flows intead.

**Comment by hadilq** [comment]
Thanks @fergusonm, I totally agree with you.
